go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,279 @@
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package mpt
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/Fantom-foundation/Carmen/go/common/interrupt"
+	"github.com/Fantom-foundation/Carmen/go/common/witness"
+	"golang.org/x/exp/maps"
+	"log"
+)
+
+// VerifyProofArchiveTrie verifies the consistency of witness proofs for an archive trie.
+// It reads the trie for each block of the archive and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots of each block.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func VerifyProofArchiveTrie(ctx context.Context, dir string, config MptConfig, from, to int, observer VerificationObserver) error {
+	trie, err := OpenArchiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	return errors.Join(
+		verifyProofArchiveTrie(ctx, trie, from, to, observer),
+		trie.Close())
+}
+
+// verifyProofArchiveTrie verifies the consistency of witness proofs for an archive trie.
+// It reads the trie for each block of the archive and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots of each block.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func verifyProofArchiveTrie(ctx context.Context, trie *ArchiveTrie, from, to int, observer VerificationObserver) error {
+	observer.StartVerification()
+	blockHeight, empty, err := trie.GetBlockHeight()
+	if err != nil {
+		observer.EndVerification(err)
+		return err
+	}
+	if empty {
+		return nil
+	}
+
+	if from < 0 {
+		from = 0
+		log.Printf("adjusting to a minimum block height: %d", from)
+	}
+
+	if to <= 0 || to > int(blockHeight) {
+		to = int(blockHeight) + 1
+		log.Printf("adjusting to a maximum block height: %d", to)
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total %d blocks", to-from))
+	for i := uint64(from); i < uint64(to); i++ {
+		observer.Progress(fmt.Sprintf("Collecting state for block... \t %d", i))
+		// collect all accounts for this block
+		visitor := accountCollectingVisitor{accounts: accountsMap{}, ctx: ctx}
+		if err := trie.VisitTrie(i, &visitor); err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		observer.Progress(fmt.Sprintf("Collecting proof for block.. \t %d \t Accounts: %d", i, len(visitor.accounts)))
+		// collect all proofs for this block
+		rawProof := make(map[string]struct{})
+		for addr, data := range visitor.accounts {
+			if interrupt.IsCancelled(ctx) {
+				return interrupt.ErrCanceled
+			}
+
+			accountProof, err := trie.CreateWitnessProof(i, addr, maps.Keys(data.storage)...)
```

Comment:
The number of storage slots of an account could be huge for some accounts. I would suggest to batch it in groups of up to 10 keys per request.

Reply:
good idea, reworked this way, now it also works better when pressing ctrl+c as previously it could take even an hour when it was processing an account with many slots.